File: projects/mini-claude-code-golang/tools.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

// 这两个结构体对应工具输入的 JSON。
type pathInput struct {
	Path string `json:"path"`
}

type writeInput struct {
	Path    string `json:"path"`
	Content string `json:"content"`
}

// callTool 负责按名字分发工具。
func callTool(root, name, raw string) string {
	switch name {
	case "listFiles":
		return listFiles(root, raw)
	case "readFile":
		return readFile(root, raw)
	case "writeFile":
		return writeFile(root, raw)
	default:
		return "未知工具: " + name
	}
}

// listFiles 列出目录内容。
func listFiles(root, raw string) string {
	in := pathInput{Path: "."}
	if strings.TrimSpace(raw) != "" && json.Unmarshal([]byte(raw), &in) != nil {
		return `listFiles 参数应为 {"path":"."}`
	}

	target, show, err := safePath(root, in.Path)
	if err != nil {
		return err.Error()
	}

	entries, err := os.ReadDir(target)
	if err != nil {
		return err.Error()
	}
	if len(entries) == 0 {
		return show + " 是空目录"
	}

	lines := []string{"目录 " + show + ":"}
	for _, entry := range entries {
		name := entry.Name()
		if entry.IsDir() {
			name += "/"
		}
		lines = append(lines, name)
	}
	return strings.Join(lines, "\n")
}

// readFile 读取文件内容；太大时只截前面一部分。
func readFile(root, raw string) string {
	var in pathInput
	if json.Unmarshal([]byte(raw), &in) != nil || strings.TrimSpace(in.Path) == "" {
		return `readFile 参数应为 {"path":"main.go"}`
	}

	target, show, err := safePath(root, in.Path)
	if err != nil {
		return err.Error()
	}

	data, err := os.ReadFile(target)
	if err != nil {
		return err.Error()
	}
	if len(data) > 8000 {
		return "文件 " + show + ":\n" + string(data[:8000]) + "\n...[已截断]"
	}
	return "文件 " + show + ":\n" + string(data)
}

// writeFile 写入文件；父目录不存在时顺手创建。
func writeFile(root, raw string) string {
	var in writeInput
	if json.Unmarshal([]byte(raw), &in) != nil || strings.TrimSpace(in.Path) == "" {
		return `writeFile 参数应为 {"path":"a.txt","content":"hello"}`
	}

	target, show, err := safePath(root, in.Path)
	if err != nil {
		return err.Error()
	}

	if err := os.MkdirAll(filepath.Dir(target), 0o755); err != nil {
		return err.Error()
	}
	if err := os.WriteFile(target, []byte(in.Content), 0o644); err != nil {
		return err.Error()
	}
	return "已写入 " + show
}

// safePath 保证工具只能访问当前工作目录。
func safePath(root, p string) (string, string, error) {
	if strings.TrimSpace(p) == "" {
		p = "."
	}

	rootAbs, err := filepath.Abs(root)
	if err != nil {
		return "", "", err
	}
	full, err := filepath.Abs(filepath.Join(rootAbs, p))
	if err != nil {
		return "", "", err
	}

	rel, err := filepath.Rel(rootAbs, full)
	if err != nil {
		return "", "", err
	}
	if rel == ".." || strings.HasPrefix(rel, ".."+string(os.PathSeparator)) {
		return "", "", fmt.Errorf("路径超出工作目录: %s", p)
	}
	if rel == "." {
		return full, ".", nil
	}
	return full, rel, nil
}

File: projects/mini-claude-code-rust/src/tools.rs
use serde::Deserialize;
use std::fs;
use std::path::{Component, Path, PathBuf};

#[derive(Deserialize)]
struct PathInput {
    path: String,
}

#[derive(Deserialize)]
struct WriteInput {
    path: String,
    content: String,
}

pub fn call_tool(root: &Path, name: &str, raw: &str) -> String {
    match name {
        "listFiles" => list_files(root, raw),
        "readFile" => read_file(root, raw),
        "writeFile" => write_file(root, raw),
        _ => format!("未知工具: {name}"),
    }
}

fn list_files(root: &Path, raw: &str) -> String {
    let input = if raw.trim().is_empty() {
        PathInput {
            path: ".".to_string(),
        }
    } else {
        match serde_json::from_str::<PathInput>(raw) {
            Ok(value) => value,
            Err(_) => return r#"listFiles 参数应为 {"path":"."}"#.to_string(),
        }
    };

    let (target, show) = match safe_path(root, &input.path) {
        Ok(value) => value,
        Err(err) => return err,
    };

    let entries = match fs::read_dir(&target) {
        Ok(value) => value,
        Err(err) => return err.to_string(),
    };

    let mut lines = vec![format!("目录 {show}:")];
    for entry in entries {
        let entry = match entry {
            Ok(value) => value,
            Err(err) => return err.to_string(),
        };

        let mut name = entry.file_name().to_string_lossy().to_string();
        if entry.path().is_dir() {
            name.push('/');
        }
        lines.push(name);
    }

    if lines.len() == 1 {
        return format!("{show} 是空目录");
    }

    lines.join("\n")
}

fn read_file(root: &Path, raw: &str) -> String {
    let input = match serde_json::from_str::<PathInput>(raw) {
        Ok(value) if !value.path.trim().is_empty() => value,
        _ => return r#"readFile 参数应为 {"path":"src/main.rs"}"#.to_string(),
    };

    let (target, show) = match safe_path(root, &input.path) {
        Ok(value) => value,
        Err(err) => return err,
    };

    let data = match fs::read(&target) {
        Ok(value) => value,
        Err(err) => return err.to_string(),
    };

    let text = String::from_utf8_lossy(&data);
    if text.chars().count() > 8_000 {
        let preview: String = text.chars().take(8_000).collect();
        return format!("文件 {show}:\n{preview}\n...[已截断]");
    }

    format!("文件 {show}:\n{text}")
}

fn write_file(root: &Path, raw: &str) -> String {
    let input = match serde_json::from_str::<WriteInput>(raw) {
        Ok(value) if !value.path.trim().is_empty() => value,
        _ => return r#"writeFile 参数应为 {"path":"a.txt","content":"hello"}"#.to_string(),
    };

    let (target, show) = match safe_path(root, &input.path) {
        Ok(value) => value,
        Err(err) => return err,
    };

    if let Some(parent) = target.parent() {
        if let Err(err) = fs::create_dir_all(parent) {
            return err.to_string();
        }
    }

    if let Err(err) = fs::write(&target, input.content) {
        return err.to_string();
    }

    format!("已写入 {show}")
}

fn safe_path(root: &Path, input: &str) -> Result<(PathBuf, String), String> {
    let root = root.canonicalize().map_err(|err| err.to_string())?;
    let raw = if input.trim().is_empty() {
        "."
    } else {
        input.trim()
    };

    if Path::new(raw).is_absolute() {
        return Err("只允许相对路径".to_string());
    }

    let mut full = root.clone();
    for component in Path::new(raw).components() {
        match component {
            Component::CurDir => {}
            Component::Normal(part) => full.push(part),
            Component::ParentDir => {
                if full == root {
                    return Err(format!("路径超出工作目录: {raw}"));
                }
                full.pop();
            }
            _ => return Err(format!("非法路径: {raw}")),
        }
    }

    let show = full
        .strip_prefix(&root)
        .ok()
        .and_then(|path| path.to_str())
        .unwrap_or(".")
        .to_string();

    if show.is_empty() {
        Ok((full, ".".to_string()))
    } else {
        Ok((full, show))
    }
}
